src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -1617,16 +1654,31 @@ void MacroAssembler::reinit_heapbase() {
   }
 }
 
-void MacroAssembler::movptr(Register Rd, address addr, int32_t &offset) {
-  int64_t imm64 = (int64_t)addr;
+void MacroAssembler::movptr(Register Rd, address addr, Register temp) {
+  int offset = 0;
+  movptr(Rd, addr, offset, temp);
+  addi(Rd, Rd, offset);
+}
+
+void MacroAssembler::movptr(Register Rd, address addr, int32_t &offset, Register temp) {
+  uint64_t uimm64 = (uint64_t)addr;
 #ifndef PRODUCT
   {
     char buffer[64];
-    snprintf(buffer, sizeof(buffer), "0x%" PRIx64, imm64);
+    snprintf(buffer, sizeof(buffer), "0x%" PRIx64, uimm64);
     block_comment(buffer);
   }
 #endif
-  assert((uintptr_t)imm64 < (1ull << 48), "48-bit overflow in address constant");
+  assert(uimm64 < (1ull << 48), "48-bit overflow in address constant");
+
+  if (temp == noreg) {
+    movptr1(Rd, uimm64, offset);
+  } else {
+    movptr2(Rd, uimm64, offset, temp);
+  }
+}
+
+void MacroAssembler::movptr1(Register Rd, uint64_t imm64, int32_t &offset) {
```

Comment:
Original code of MacroAssembler:: movptr(...) is bit tricky at `upper -= lower;` to understand for me, and I think new MacroAssembler:: movptr2(...) uses the similar way at `lower30 - low12`.
I can add some comment later to help future understanding.

Reply:
Yes, it's very confusing with 12 bit *signed*, imm 11 bits + 1 signed bit.
Therefore we need to use arithmetic shift to create "12 bit sign value".
(hence why they do not use 12ull, to signal this is an arithmetic shift)

So if we need to set bit 12 this value it will be negative plus the bit pattern we want in the other 11 bits.
To compensate for that we need to lui() a larger value, and that value would be low30 - low12.
If low12 is postive it's just removing those bits, otherwise we also need to add 4096 (bit 12).
Hence low30 - -low12 would instead add to low30, the mid18 can thus be extended to 19 bits.

The 20 bit is sign bit so we must stay away from it otherwise lui() will sign extend the 19+1 imm value.

I don't think that helped, but I tried at least :)

Reply:
Thanks for explanation. It also took me a while to figure it out.

Then I will try to add some comment for it later. Maybe also try to refactor movptr2() if I can find out a way to make it more clear.